Fix quick filename search (^S):
- Avoid matching partial multibyte character.
  E.g. a quick search for fooá when the only file beginning with foo is fooé
  don't match the first byte of é.
- Backspace remove a whole multibyte character, not just one byte.

diff -Naurdp mc-4.6.1.orig/src/screen.c mc-4.6.1/src/screen.c
--- mc-4.6.1.orig/src/screen.c	2005-09-05 17:08:08.000000000 +0200
+++ mc-4.6.1/src/screen.c	2005-09-05 17:23:00.000000000 +0200
@@ -2000,11 +2000,24 @@ do_search (WPanel *panel, int c_code)
     int i;
     int wrapped = 0;
     int found;
+    int prevpos, pos;
+    int j;
+    mbstate_t mbs;
 
     l = strlen (panel->search_buffer);
     if (c_code == KEY_BACKSPACE) {
-	if (l)
-	    panel->search_buffer[--l] = '\0';
+	if (l) {
+	    prevpos = pos = 0;
+	    memset (&mbs, 0, sizeof (mbs));
+	    while (pos < l) {
+		prevpos = pos;
+		j = mbrlen (panel->search_buffer + pos, l - pos, &mbs);
+		if (j <= 0) break;
+		pos += j;
+	    }
+	    --l;
+	    panel->search_buffer[prevpos] = 0;
+	}
     } else {
 	if (c_code && l < sizeof (panel->search_buffer)) {
 	    panel->search_buffer[l] = c_code;
@@ -2013,6 +2026,14 @@ do_search (WPanel *panel, int c_code)
 	}
     }
 
+    prevpos = pos = 0;
+    memset (&mbs, 0, sizeof (mbs));
+    while (pos < l) {
+	prevpos = pos;
+	j = mbrlen (panel->search_buffer + pos, l - pos, &mbs);
+	if (j <= 0) break;
+	pos += j;
+    }
     found = 0;
     for (i = panel->selected; !wrapped || i != panel->selected; i++) {
 	if (i >= panel->count) {
@@ -2023,9 +2044,9 @@ do_search (WPanel *panel, int c_code)
 	}
 	if (panel->
 	    case_sensitive
-	    ? (strncmp (panel->dir.list[i].fname, panel->search_buffer, l)
+	    ? (strncmp (panel->dir.list[i].fname, panel->search_buffer, pos)
 	       == 0) : (g_strncasecmp (panel->dir.list[i].fname,
-				       panel->search_buffer, l) == 0)) {
+				       panel->search_buffer, pos) == 0)) {
 	    unselect_item (panel);
 	    panel->selected = i;
 	    select_item (panel);
@@ -2034,7 +2055,7 @@ do_search (WPanel *panel, int c_code)
 	}
     }
     if (!found)
-	panel->search_buffer[--l] = 0;
+	panel->search_buffer[prevpos] = 0;
 
     paint_panel (panel);
 }
